scheme.py

```diff
@@ -1,62 +1,187 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
-from typing import List, Optional, Tuple, Iterable
+from typing import (
+    List, Optional, Tuple, Iterable, Sequence, Union, cast, TYPE_CHECKING)
+
+if TYPE_CHECKING:
+    from environment import Environment
 
 
-@dataclass
 class SExp:
     """An s-expression base class"""
-    pass
+    def __init__(self):
+        self._environment: Optional[Environment] = None
+
+    @property
+    def environment(self) -> Environment:
+        if self._environment is None:
+            raise Exception('Environment not set')
 
+        return self._environment
 
-@dataclass(frozen=True, order=True)
+    @environment.setter
+    def environment(self, env: Environment):
+        self._environment = env
+
+
+@dataclass(order=True)
 class SNum(SExp):
     """A lisp number"""
     value: int
 
     def __str__(self):
         return str(self.value)
 
+    def __hash__(self):
+        return hash(self.value)
```

Comment:
Do we intentionally want items with different environments to hash the same? I suppose the notion here is "immutable except for setting the environment"? I assume `frozen=True` breaks that assignment?

Reply:
For symbols, I think the answer is definitely yes. For numbers, do we even need them to be hashable at all? Symbols need to be hashable for environment lookup and storage, but I don't think numbers do.